torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -415,39 +439,41 @@ void RRefContext::addSelfAsFork(c10::intrusive_ptr<OwnerRRef>& rref) {
 void RRefContext::addForkOfOwner(const RRefId& rrefId, const ForkId& forkId) {
   std::lock_guard<std::mutex> lock(mutex_);
   auto& rrefForks = forks_[rrefId];
-  TORCH_INTERNAL_ASSERT(
-      rrefForks.find(forkId) == rrefForks.end(),
-      "Got fork notification twice on the same RRef ",
-      forkId);
-  rrefForks.insert(forkId);
+  // We first check whether the child exists in rrefForks. It's possible
+  // the child may have been added by a previous send attempt, and this check
+  // (as opposed to an assertion here) ensures that messages that trigger this
+  // function are idempotent.
+  if (rrefForks.find(forkId) == rrefForks.end()) {
+    rrefForks.insert(forkId);
+  }
 }
 
 c10::intrusive_ptr<RRef> RRefContext::delForkOfOwner(
     const RRefId& rrefId,
     const ForkId& forkId) {
   c10::intrusive_ptr<RRef> deletedRRef;
+  // There were previously multiple TORCH_CHECKs in this function that checked
+  // whether the passed in fork was known by the user and whether the fork had
+  // already been deleted. These assertions are now replaced with nested if
+  // statements to ensure this function is idempotent. This makes it safe to
+  // retry RRefUserDelete messages.
   {
     std::lock_guard<std::mutex> lock(mutex_);
     auto rrefIter = forks_.find(rrefId);
-    TORCH_INTERNAL_ASSERT(
-        rrefIter != forks_.end(),
-        "Inconsistent states, deleting a fork before the owner knows it.");
-    auto& rrefForks = rrefIter->second;
-    auto forkIter = rrefForks.find(forkId);
-    TORCH_INTERNAL_ASSERT(
-        forkIter != rrefForks.end(),
-        "Attempt to delete a non-exist fork ",
-        forkId);
-
-    rrefForks.erase(forkId);
-
-    if (rrefForks.empty()) {
-      auto ownerIter = owners_.find(rrefId);
-      if (ownerIter != owners_.end()) {
-        deletedRRef = ownerIter->second;
-        owners_.erase(ownerIter);
+    if (rrefIter != forks_.end()) {
```

Comment:
Can we add a LOG(INFO) with details of the rref (rref id and fork id) in the case that it was not found? Would be goo d for debugging purposes later on.